### PR TITLE
Generate user login state from access lists and integrate into certificates.

### PR DESCRIPTION
**Pull request by mdwn**
On login, the user login state will be generated, using access lists to register additional roles and traits that will be inserted into the user's certificate.

Tests have been added to exercise this as well.

[Relevant section in the RFD](https://github.com/gravitational/teleport.e/blob/master/rfd/0006e-access-lists.md#userloginstate-object).

Small low key benchmark:

```
With login state

AuthenticateUser begin: 2023-07-20 13:35:22.553049 -0400 EDT m=+23.786587251
AuthenticateUser end: 2023-07-20 13:35:22.66828 -0400 EDT m=+23.901817918
AuthenticateUser delta: 115.230667ms
GenerateCert begin: 2023-07-20 13:35:22.668675 -0400 EDT m=+23.902212501
GenerateCert end: 2023-07-20 13:35:22.673191 -0400 EDT m=+23.906728626
GenerateCert delta: 4.516125ms

Without login state

AuthenticateUser begin: 2023-07-20 13:40:13.515064 -0400 EDT m=+12.754441292
AuthenticateUser end: 2023-07-20 13:40:13.636613 -0400 EDT m=+12.875989501
AuthenticateUser delta: 121.548209ms
GenerateCert begin: 2023-07-20 13:40:13.63706 -0400 EDT m=+12.876436501
GenerateCert end: 2023-07-20 13:40:13.641253 -0400 EDT m=+12.880629459
GenerateCert delta: 4.192958ms

With login state + 2 access lists (not a member)

AuthenticateUser begin: 2023-07-20 13:50:17.48219 -0400 EDT m=+126.730266376
AuthenticateUser end: 2023-07-20 13:50:17.595715 -0400 EDT m=+126.843790251
AuthenticateUser delta: 113.523875ms
GenerateCert begin: 2023-07-20 13:50:17.596153 -0400 EDT m=+126.844228043
GenerateCert end: 2023-07-20 13:50:17.60055 -0400 EDT m=+126.848625126
GenerateCert delta: 4.397083ms

With login state + 2 access lists (member)

AuthenticateUser begin: 2023-07-20 14:05:50.527176 -0400 EDT m=+141.383355001
AuthenticateUser end: 2023-07-20 14:05:50.633066 -0400 EDT m=+141.489245043
AuthenticateUser delta: 105.890042ms
GenerateCert begin: 2023-07-20 14:05:50.633598 -0400 EDT m=+141.489777001
GenerateCert end: 2023-07-20 14:05:50.637755 -0400 EDT m=+141.493933918
GenerateCert delta: 4.156917ms
```

### CODE REVIEW

File: lib/auth/auth.go [hunk 1]
```diff
@@ -245,6 +246,12 @@ func NewServer(cfg *InitConfig, opts ...ServerOption) (*Server, error) {
 	if cfg.UserPreferences == nil {
 		cfg.UserPreferences = local.NewUserPreferencesService(cfg.Backend)
 	}
+	if cfg.UserLoginState == nil {
+		cfg.UserLoginState, err = local.NewUserLoginStateService(cfg.Backend)
```
Comment (mdwn) [inline lib/auth/auth.go:250]: Should this be a memory backend here?

Comment (mdwn) [reply lib/auth/auth.go]: I've migrated this to use a memory backend so that it can be quickly cached/looked up elsewhere.

File: lib/auth/userloginstate/generator.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+	// Create an identity for testing membership to access lists.
+	identity := tlsca.Identity{
+		Username: user.GetName(),
+		Groups:   user.GetRoles(),
+		Traits:   user.GetTraits(),
+	}
+
+	for _, accessList := range accessLists {
+		if err := services.IsMember(identity, g.clock, accessList); err != nil {
+			continue
```
Comment (Tener) [inline lib/auth/userloginstate/generator.go:126]: The order in which we process elements of `accessLists` can affect the end result of the function here. Specifically, we can have a situation where access list `foo` will effectively grant access for `bar`, but only if it is processed first. 

I think we can fix this fairly easily: if user was found to be a member of any of the lists, retry the whole loop again (but only for rejected items).

This can blow up spectacularly with large or malicious access lists, but we can try to optimize it as needed.

**Comment by Tener** [comment]
@mdwn is there an RFD or parent issue I can read? I feel like I'm missing the context here.

File: lib/auth/userloginstate/generator.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+	for k, v := range state.Spec.Traits {
+		state.Spec.Traits[k] = utils.Deduplicate(v)
+	}
+
+	// Remove roles that don't exist in the backend so that we don't generate certs for non-existent roles.
+	// Doing so can prevent login from working properly. This could occur if access lists refer to roles that
+	// no longer exist, for example.
+	existingRoles := []string{}
+	for _, role := range state.Spec.Roles {
+		if _, err := g.roles.GetRole(ctx, role); err == nil {
```
Comment (Tener) [inline lib/auth/userloginstate/generator.go:153]: Can we just change the available method to `GetRoles`? I think this would be faster by virtue of fewer RPC calls. We can get away from checking for `trace.IsNotFound` err, which I find imprecise at best.

File: lib/auth/userloginstate/generator_test.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+func newAccessList(t *testing.T, name string, members []string, roles []string, traits trait.Traits) *accesslist.AccessList {
+	t.Helper()
+
+	alMembers := make([]accesslist.Member, len(members))
+	for i, member := range members {
+		alMembers[i] = accesslist.Member{
+			Name:    member,
+			Joined:  time.Now(),
+			Expires: time.Now().Add(24 * time.Hour),
```
Comment (Tener) [inline lib/auth/userloginstate/generator_test.go:205]: I wonder if we should consider the expiry time when issuing certificates. Otherwise, you can issue a certificate 5 minutes before someone's membership expires, but they can keep using it for as long as the certificate is valid: potentially a long time.

File: lib/auth/userloginstate/generator_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+func newAccessList(t *testing.T, name string, members []string, roles []string, traits trait.Traits) *accesslist.AccessList {
+	t.Helper()
+
+	alMembers := make([]accesslist.Member, len(members))
+	for i, member := range members {
+		alMembers[i] = accesslist.Member{
+			Name:    member,
+			Joined:  time.Now(),
```
Comment (Tener) [inline lib/auth/userloginstate/generator_test.go:204]: This is strictly outside of scope of this particular PR, but I wonder if we could make the access lists more flexible. Right now we have:

```
// IsMember will return true if the user is a member for the current list.
func IsMember(identity tlsca.Identity, clock clockwork.Clock, accessList *accesslist.AccessList) error {
	username := identity.Username
	for _, member := range accessList.Spec.Members {
		if member.Name == username && clock.Now().Before(member.Expires) {
			if !userMeetsRequirements(identity, accessList.Spec.MembershipRequires) {
				return trace.AccessDenied("user %s does not meet membership requirements", username)
			}
			return nil
		}
	}

	return trace.NotFound("user %s is not a member of the access list", username)
}
```

If a person is a member of the access list, they are on it until their membership expires. And this is it. 

We could make the membership more flexible, though: for example, based on time of day or day of week, or any other flexible schedule. In the simplest case you should be able to issue membership in the future and have it only valid after given point of time.

File: lib/auth/userloginstate/generator_test.go (hunk 4 above)
Comment (mdwn) [reply lib/auth/userloginstate/generator_test.go:205]: IIRC we've actually got this issue as with access requests and other things -- I'm unsure of the right thing to do here either. I'd love to get more opinions, perhaps @espadolini?

File: lib/auth/userloginstate/generator_test.go (hunk 5 above)
Comment (mdwn) [reply lib/auth/userloginstate/generator_test.go:204]: Interesting. That's definitely something we can look into in the future! It would make for some interesting use cases.

Comment (Tener) [reply lib/auth/userloginstate/generator_test.go:204]: Yep, shift-based entitlements come to mind in particular.

File: lib/auth/auth.go [hunk 6]
```diff
@@ -2333,20 +2363,42 @@ func generateCert(a *Server, req certRequest, caType types.CertAuthType) (*proto
...
+		roles = append(roles, uls.GetRoles()...)
+
+		// Copy traits to the traits from the request.
+		if len(uls.GetTraits()) > 0 {
+			if traits == nil {
+				traits = map[string][]string{}
+			}
+			for k, v := range uls.GetTraits() {
+				traits[k] = append(traits[k], v...)
+			}
```
[the remark this answers is not in the document]
Comment (mdwn) [reply lib/auth/auth.go:2382]: I've deduplicated and added a test here. Will post an update to the PR shortly.

File: lib/auth/userloginstate/generator.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+		roleLookup[role.GetName()] = true
+	}
+
+	existingRoles := []string{}
+	for _, role := range state.Spec.Roles {
+		if roleLookup[role] {
+			existingRoles = append(existingRoles, role)
+		}
+	}
+	state.Spec.Roles = existingRoles
```
Comment (Tener) [inline lib/auth/userloginstate/generator.go:172]: I think we should warn that given roles were granted but do not exist. This may be due to the role being removed, but may well be due to a typo in the definition. Having a warning here is a step toward finding out about that typo.

File: lib/auth/userloginstate/generator.go (hunk 2 above)
Comment (mdwn) [reply lib/auth/userloginstate/generator.go:126]: So I don't think this is actually the case here -- access list membership requirements is dependent on the user's fixed definition, not the emerging definition.

Comment (Tener) [reply lib/auth/userloginstate/generator.go:126]: Hmm, I'll dig into this and check my theory.

Comment (Tener) [reply lib/auth/userloginstate/generator.go:126]: Ok, I see I was wrong with this. This does simplify things, although I wonder if we will ever come to change this.

**Review by fspmarshall — COMMENTED** [review]
As I understood the original discussion about this feature, user login state was intended to be a complete representation of the "post-mapping" user-state that we could use as a drop-in replacement for the user resource as-appropriate. As currently implemented, the various cert-generation methods are still loading the user resource as the primary source of truth, and then using the login state to extend that.  So currently we end up with a pattern like:

```golang
	user, _ := a.GetUser(req.Username, false)
	// ...
	accessInfo := services.AccessInfoFromUser(user)
	// ...
	checker, _ := services.NewAccessChecker(accessInfo, ...)
	// ...
	traits, roles = addUserLoginStateToTraitsAndRoles(uls, req.traits, req.checker.RoleNames())
	// ...etc
```

Which means that we're still mixing pre-login and post-login state (e.g. the access checker is initialized with your pre-mapping state).

I don't think this is how we should be going about it.  I think we'd be better off migrating to the login state as the new source of truth, and tweaking things like `AccessChecker` to work with an interface that abstracts over the different between user state and user.  Then we could do something like this (oversimplified):

```golang
func getUserOrLoginState(name string) UserStateInterface {
    if uls, _ := getUserLoginState(name); uls != nil {
        return uls
    }
    return getUser(name)
}

// ...

accessInfo := services.AccessInfoFromUserStateInterface(getUserOrLoginState())

// remaining logic is totally identical to what we had before
```

Basically, I feel like it's going to cause issues if we sometimes work from the pre-mapping state and sometimes from the post-mapping state.  Much better to work exclusively with the post-mapping state, and fall back to the pre-mapping state of the post mapping state does not exist (e.g. due to login having been handled by an older auth server).

WDYT?  Does this sound reasonable/doable?

**Comment by mdwn** [comment]
> As I understood the original discussion about this feature, user login state was intended to be a complete representation of the "post-mapping" user-state that we could use as a drop-in replacement for the user resource as-appropriate. As currently implemented, the various cert-generation methods are still loading the user resource as the primary source of truth, and then using the login state to extend that. So currently we end up with a pattern like:
> 
> ```go
> 	user, _ := a.GetUser(req.Username, false)
> 	// ...
> 	accessInfo := services.AccessInfoFromUser(user)
> 	// ...
> 	checker, _ := services.NewAccessChecker(accessInfo, ...)
> 	// ...
> 	traits, roles = addUserLoginStateToTraitsAndRoles(uls, req.traits, req.checker.RoleNames())
> 	// ...etc
> ```
> 
> Which means that we're still mixing pre-login and post-login state (e.g. the access checker is initialized with your pre-mapping state).
> 
> I don't think this is how we should be going about it. I think we'd be better off migrating to the login state as the new source of truth, and tweaking things like `AccessChecker` to work with an interface that abstracts over the different between user state and user. Then we could do something like this (oversimplified):
> 
> ```go
> func getUserOrLoginState(name string) UserStateInterface {
>     if uls, _ := getUserLoginState(name); uls != nil {
>         return uls
>     }
>     return getUser(name)
> }
> 
> // ...
> 
> accessInfo := services.AccessInfoFromUserStateInterface(getUserOrLoginState())
> 
> // remaining logic is totally identical to what we had before
> ```
> 
> Basically, I feel like it's going to cause issues if we sometimes work from the pre-mapping state and sometimes from the post-mapping state. Much better to work exclusively with the post-mapping state, and fall back to the pre-mapping state of the post mapping state does not exist (e.g. due to login having been handled by an older auth server).
> 
> WDYT? Does this sound reasonable/doable?

I think this sounds pretty reasonable. I don't think it'll be significantly more work than what I've already got here, it'll just involve starting with the "user" as the base user login state and then going from there. I'll update when I've made these changes.

File: lib/auth/auth.go [hunk 8]
```diff
@@ -1666,6 +1698,15 @@ func certRequestDeviceExtensions(ext tlsca.DeviceExtensions) certRequestOption {
...
 }
 
+// getUserOrLoginState will return the given user or the login state associated with the user.
+func (a *Server) getUserOrLoginState(ctx context.Context, user types.User) services.UserState {
+	if uls, _ := a.GetUserLoginState(ctx, user.GetName()); uls != nil {
+		return uls
+	}
+
+	return user
+}
```
Comment (fspmarshall) [inline lib/auth/auth.go:1707]: IMO we shouldn't be loading the user resource at all for access-control logic unless we explicitly get a not found error, it will be too easy to accidentally use the pre-mapping state in cases where it would be wrong to do so (which are most cases now... user login state effectively replaces `types.User` for almost everything in `auth_with_roles.go`).

Off the top of my head, I'm fairly sure only cases where we should ever have to load the legacy user state for access-control decisions should be either:

1. We just upgraded and there are users still logged in from before the login state feature rolled out.
2. The identity in question is a bot or other piece of automation whose certs were created directly rather than being the result of a normal "login" process.

Based on that, I'd suggest reworking this function to something like this:

```golang
// getUserOrLoginState will return login state for a user if available, falling back to returning the user resource.
func (a *Server) getUserOrLoginState(ctx context.Context, username string) (services.UserState, error) {
	uls, err := a.GetUserLoginState(ctx, username)
	if err != nil && !trace.IsNotFound(err) {
		return nil, trace.Wrap(err)
	}
	
	if err == nil {
		return uls, nil
	}

	user, err := a.GetUser(username,  false)
	return user, trace.Wrap(err)
}
```

File: lib/auth/userloginstate/generator_test.go (hunk 4 above)
Comment (espadolini) [reply lib/auth/userloginstate/generator_test.go:230]: For access requests we've only recently added an explicit flag to let them be issued for longer than the original credentials; in general, the time range in which anything derived from an identity is valid should be contained in the original identity's validity unless otherwise explicitly specified, IMO.

File: lib/auth/auth.go (hunk 8 above)
Comment (mdwn) [reply lib/auth/auth.go]: I've implemented this change and a number of the downstream impacts here. I'll resolve this, but I wouldn't be surprised if there were some additional issues.

**Comment by mdwn** [comment]
@Tener @espadolini @fspmarshall This should be good to review now, I've worked through the kinks here.

File: lib/auth/auth.go [hunk 9]
```diff
@@ -1629,6 +1650,21 @@ func certRequestDeviceExtensions(ext tlsca.DeviceExtensions) certRequestOption {
 	}
 }
 
+// getUserOrLoginState will return the given user or the login state associated with the user.
```
Comment (Tener) [inline lib/auth/auth.go:1653]: Is there a legitimate case for the user *not* to have user logins state?

If not, we could force the user to re-login by deleting the user login state.

File: lib/auth/userloginstate/generator_test.go (hunk 4 above)
Comment (mdwn) [reply lib/auth/userloginstate/generator_test.go:230]: Would you be amenable to addressing this in a follow on? I'm a little worried about making this PR more complicated than it already is.

**Review by fspmarshall — APPROVED** [review]

File: lib/auth/auth.go (hunk 9 above)
Comment (fspmarshall) [reply lib/auth/auth.go:1653]: Certificates issued directly with `tctl auth sign` (e.g. certs for a bot) won't have an associated login state.  There may be other cases as well.

File: lib/auth/userloginstate/generator_test.go (hunk 4 above)
Comment (Tener) [reply lib/auth/userloginstate/generator_test.go:230]: Fair enough, I think such changes would naturally apply on top of the current patchset.

cc @jentfoo for awareness, this issue may benefit from deeper thought or a follow-up audit.

**Review by Tener — APPROVED** [review]

File: lib/auth/auth.go
Comment (ravicious) [inline lib/auth/auth.go:1]: git bisect tells me that this introduced a regression in Connect My Computer that was not caught by existing tests and I don't understand why.

In general, when you start Connect My Computer setup (instructions in the description of #30905), Connect creates a new role in the cluster (`CreateConnectMyComputerRole` RPC) and assigns it to the current user. At this point, we have to refresh the role list in local certs. To achieve this, we do this:

https://github.com/gravitational/teleport/blob/0ed045aa001260fb3ad740f65068f86e5f852fef/lib/teleterm/services/connectmycomputer/connectmycomputer.go#L181-L199

This used to update the certs on disk with the current role list from the backend. It's the same operation that `tsh request drop` does when you tell it to drop a certain access request.

After this is done, the Electron app calls tshd's `GetCluster` RPC which includes the role list read from the cert on disk. I noticed that it now doesn't get updated and git bisect led me to this PR. `tsh status` run from inside Connect also doesn't return the new role.

What's weird is that there's an integration test which calls `CreateConnectMyComputerRole`. It specifically checks if the certs on disk include the new role:

https://github.com/gravitational/teleport/blob/12945167b49eb372df32000c622ab2ec1674752a/integration/teleterm_test.go#L546-L555

The test passes just fine. It's the "role does not exist" test from those table tests above.

I also added a test for `GenerateUserCerts` which also passes:

https://github.com/gravitational/teleport/blob/12945167b49eb372df32000c622ab2ec1674752a/lib/auth/access_request_test.go#L633-L638

Given the changes in this PR, do you have any clue as to why this would update the cert in the integration test but not in the app?

Comment (ravicious) [reply lib/auth/auth.go:1]: I suppose this is related to the change in `lib/auth/auth_with_roles.go`.

Comment (strideynet) [reply lib/auth/auth.go:1]: We have reason to believe this may also have caused a regression for Machine ID. Looking into it.

Comment (ravicious) [reply lib/auth/auth.go:1]: #30978 doesn't fix the regression in Connect My Computer, unfortunately.

Comment (ravicious) [reply lib/auth/auth.go:1]: Ok, so I see the regression happened because Connect My Computer depended on the following interactions:

1. Log in with roles foo and bar.
2. Create a new role baz and assign it to the current user.
3. Drop a bogus access request (== refresh the role list based on the current backend state).

Now it doesn't work like this, because at point 3 the role list is not updated based on the backend state, but rather on the "cached" login state from 1.

We used the hack with dropping a bogus access request because we wanted to have a way to refresh the role list without asking the user for credentials again. This is the same behavior that `tsh request drop` needed.

However, with the login state in place, it's in fact better (more secure) for `tsh request drop` to use the login state rather than backend state. This doesn't work for Connect My Computer because we need fresh roles.

Comment (ravicious) [reply lib/auth/auth.go:1]: I suppose in the end we shouldn't consider this a regression but rather an end to a hacky workaround Connect was using to make Connect My Computer interaction free, right?

Long term, we want Connect My Computer to not depend on creating a new role for various reasons, one of them being that creating a new role and assigning it to the current user [doesn't work for SSO users](https://gravitational.slack.com/archives/C03FJA391M3/p1691594018337799?thread_ts=1691513122.092849&cid=C03FJA391M3).

In the meantime, we could force the user to log in again after the role is added. This is kind of a hit to the UX of this whole feature, but we could live with that for now.

Comment (ravicious) [reply lib/auth/auth.go:1]: One thing I still don't understand is how it slipped through the integration test, as I mentioned in my original comment. I suppose it's an issue with some kind of cache. ;)

Comment (mdwn) [reply lib/auth/auth.go:1]: Hrrrm, it definitely wasn't intentional to change this behavior. Where exactly did I change that here? IMO we should revert that line, because IMO I'd rather keep the existing UX and improve it later than have a workaround.

Comment (ravicious) [reply lib/auth/auth.go:1]: It's [the change in `lib/auth/auth_with_roles.go`](https://github.com/gravitational/teleport/pull/29364/files#diff-ec1bd6a036583633acf2fa781d0d32d4355c2f144ce484fa9a7f780b89c3311d). We used to build `accessInfo` based on the current backend state of the user, this PR made it so that it's based on the login state instead.

Reverting it back to `accessInfo = services.AccessInfoFromUser(user)` fixes the regression.

Do you have any clue as to why this wasn't caught in the integration test? Since [the integration test doesn't go through the normal login procedure](https://github.com/gravitational/teleport/blob/5f142d37588031c684a87bb5717ab4834a3a6ce5/integration/teleterm_test.go#L675-L685), I suppose login state was empty and thus `accessInfo` was calculated from the user resource, right?

Comment (mdwn) [reply lib/auth/auth.go:1]: Yep, that would make sense. I think you're right, the user state was probably just empty.

Comment (ravicious) [reply lib/auth/auth.go:1]: I made [a PR which reverts this change in `lib/auth/auth_with_roles.go`](https://github.com/gravitational/teleport/pull/31068). If you feel that we should revert it, then I say let's do it.

It'd be good to get this merged before the test plan starts, but this might be quite tight.
